### DISCUSSION TITLE
How can I set the Sentry environment for the browser at runtime?

### DISCUSSION THREAD

**Discussion by michaelschufi** [vercel/next.js#30039] [Help]
Hi everyone!

Could anyone maybe help me with the first steps of Sentry.io integration? 
I've now configured `sentry.client.config.js` and `sentry.server.config.js` but I am having trouble configuring it at runtime.

We are deploying our single build to multiple environments, building pages dynamically using Incremental Static Regeneration. However, Sentry seems to ignore the environment variable at runtime (defined in `sentry.client.config.js`). My guess is, that `Sentry.init(...)`  is only called during the build and never at runtime.

Now to my question: How can I set the sentry environment for the browser at runtime? 

My current sentry client config:

```
import * as Sentry from '@sentry/nextjs';
import getConfig from 'next/config';

const { publicRuntimeConfig } = getConfig();

const SENTRY_DSN = process.env.SENTRY_DSN || process.env.NEXT_PUBLIC_SENTRY_DSN;

Sentry.init({
  dsn:
    SENTRY_DSN ||
    'https://foo@bar.ingest.sentry.io/1234567',
  tracesSampleRate: process.env.SENTRY_TRACES_SAMPLE_RATE || 0,
  environment: publicRuntimeConfig.SENTRY_ENVIRONMENT || 'development',
});
```

and in `next.config.js`:

```
const moduleExports = withBundleAnalyzer({
  publicRuntimeConfig: {
    SENTRY_ENVIRONMENT: process.env.SENTRY_ENVIRONMENT || 'development',
  },
});

// ...

module.exports = isSentryEnabled
  ? withSentryConfig(moduleExports, SentryWebpackPluginOptions)
  : moduleExports;
```

How can I set the sentry environment for the browser at runtime if not with `publicRuntimeConfig`?

Any help would be greatly appreciated!

Thanks in advance!
Michael

**Comment by Marcosld** [comment]
We actually can't migrate our project to app/ folder because of this :(

**Reply by jylertones** [reply]
A couple years later, did you ever figure this out?

**Reply by Marcosld** [reply]
We are running a script before the server starts up that replaces some placeholders in the javascript bundles to the real sentry URLs 😅 . Best approach we could figure.

**Reply by nickv2002** [reply]
@Marcosld that sounds exactly what we need. Could you share or elaborate on which Sentry values and/or share a sanitized version of the startup script?

**Reply by Marcosld** [reply]
Of course!

This is the `entrypoint.sh` that we use in our Dockerfile:
```bash
#!/bin/bash 

echo "Replacing sentry bundles with environment variables found..."
placeholder_sentry_dsn="<placeholder-sentry-dsn>"
placeholder_sentry_environment="<placeholder-sentry-environment>"
placeholder_sentry_release="<placeholder-sentry-release>"
placeholder_sentry_performance_sampling="<placeholder-sentry-performance-sampling>"

RELEASE_VERSION=`grep version ./package.json | awk -F \" '{print $4}'`

grep -qr "$placeholder_sentry_dsn" ./.next/static

[[ $? -ne 0 ]] && echo "Could not find sentry bundles to replace environment variables" && exit 1

find ./.next/static -type f -exec sed -i -e "s~$placeholder_sentry_dsn~$SENTRY_DSN~g" -e "s~$placeholder_sentry_performance_sampling~$SENTRY_PERFORMANCE_SAMPLING~g" -e "s~$placeholder_sentry_environment~$CONFIG_NAMESPACE~g" -e "s~$placeholder_sentry_release~$RELEASE_VERSION~g" {} \;

[[ $? -ne 0 ]] && echo "There was an error replacing environment variables" && exit 1

echo "Placeholders replaced!"

exec "$@"
```
As you can see, here we are replacing the placeholders by environment variables set on runtime.

Then we define our `sentry.client.config.ts` like this:

```typescript
import * as Sentry from '@sentry/nextjs';

const getSentryTracingSampleRate = () => (Number('<placeholder-sentry-performance-sampling>') || 0) / 100;

Sentry.init({
      dsn: '<placeholder-sentry-dsn>',
      tracesSampleRate: getSentryTracingSampleRate(),
      environment: '<placeholder-sentry-environment>',
      release: '<placeholder-sentry-release>',
  });
```
Hope it helps!!

**Reply by nickv2002** [reply]
Thanks for the great response @Marcosld! Took a while because of other work stuff but I was able to set up something similar using our env and the Bourne shell (since we don't have Bash on the Alpine base images we use):

```sh
#!/bin/sh
echo "Replacing sentry bundles with environment variables"
# This should not be needed since use the same DSN for the whole project
# placeholder_sentry_dsn="<placeholder-sentry-dsn>"
placeholder_sentry_environment="<placeholder-sentry-environment>"
placeholder_sentry_performance_sampling="<placeholder-sentry-performance-sampling>"
placeholder_sentry_release="<placeholder-sentry-release>"
# placeholder_sentry_release is the key, RELEASE_VERSION below is the value
RELEASE_VERSION=$(grep version ./package.json | awk -F \" '{print $4}')

if ! find ./.next/static -type f -exec sed -i \
  -e "s~$placeholder_sentry_performance_sampling~$SENTRY_PERFORMANCE_SAMPLING~g" \
  -e "s~$placeholder_sentry_environment~$APP_ENV~g" \
  -e "s~$placeholder_sentry_release~$RELEASE_VERSION~g" {} \;
then
    echo "There was an error replacing environment variables" && exit 1
fi

echo "Placeholders replaced!"

exec "$@"
```
(Our `APP_ENV` is a pre-existing environment variable that was easy to key into for Sentry's environment.)
